crates/ruff/src/rules/pylint/rules/cyclic_import.rs

```diff
@@ -0,0 +1,354 @@
+use std::path::Path;
+use std::sync::Arc;
+
+use log::debug;
+
+use rustc_hash::{FxHashMap, FxHashSet};
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::helpers::to_module_path;
+use ruff_python_ast::imports::{ImportMap, ModuleImport};
+
+#[violation]
+pub struct CyclicImport {
+    pub cycle: String,
+}
+
+impl Violation for CyclicImport {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Cyclic import ({}) (cyclic-import)", self.cycle)
+    }
+}
+
+struct VisitedAndCycles {
+    fully_visited: FxHashSet<Arc<str>>,
+    cycles: Option<FxHashSet<Vec<Arc<str>>>>,
+}
+
+impl VisitedAndCycles {
+    fn new(fully_visited: FxHashSet<Arc<str>>, cycles: FxHashSet<Vec<Arc<str>>>) -> Self {
+        if cycles.is_empty() {
+            Self {
+                fully_visited,
+                cycles: None,
+            }
+        } else {
+            Self {
+                fully_visited,
+                cycles: Some(cycles),
+            }
+        }
+    }
+}
+
+struct CyclicImportChecker<'a> {
+    imports: &'a FxHashMap<Arc<str>, Vec<ModuleImport>>,
+}
+
+impl CyclicImportChecker<'_> {
+    fn has_cycles(&self, name: &Arc<str>) -> VisitedAndCycles {
+        let mut fully_visited: FxHashSet<Arc<str>> = FxHashSet::default();
```

Comment:
I'm still reasonably new to this so do correct me if I'm wrong. The way I've done it I believe that we should have the same `str` potentially owned by multiple owners, though this might also not always be the case and I may indeed have a different `Arc<str>` which does contain the same value as another. For a `Box<T>` only one thing can own `T`. Cloning a `Box<T>` requires cloning the contents of the `Box` which isn't what I want.

Reply:
That's a good point! Using refcounting here does make sense to avoid allocating the same module name many times. Would it be possible to use an `Rc` instead of `Arc` or is this data structure shared across thread boundaries?

Reply:
So the input map is created in a way that is 'threaded' due to using `par_iter` which is where the first use of `Arc<str>` comes in. With the changes for referring to module names by id instead, I think I can do this with just `Rc`. I'm working on it just now, and it'll look quite different after so let me know what you think when it's done.